clippy_lints/src/utils/author.rs

```diff
@@ -190,43 +195,343 @@ impl PrintVisitor {
+            LitKind::Byte(b) => println!("    if let LitKind::Byte({}) = {}.node;", b, lit_pat),
+            LitKind::Int(i, suffix) => {
+                let int_ty = match suffix {
+                    LitIntType::Signed(int_ty) => format!("LitIntType::Signed(IntTy::{:?})", int_ty),
+                    LitIntType::Unsigned(uint_ty) => format!("LitIntType::Unsigned(UintTy::{:?})", uint_ty),
+                    LitIntType::Unsuffixed => String::from("LitIntType::Unsuffixed"),
+                };
+
+                println!("    if let LitKind::Int({}, {}) = {}.node;", i, int_ty, lit_pat);
+            },
+            LitKind::Float(_, suffix) => {
+                let float_ty = match suffix {
+                    LitFloatType::Suffixed(suffix_ty) => format!("LitFloatType::Suffixed(FloatTy::{:?})", suffix_ty),
+                    LitFloatType::Unsuffixed => String::from("LitFloatType::Unsuffixed"),
+                };
+
+                println!("    if let LitKind::Float(_, {}) = {}.node;", float_ty, lit_pat);
+            },
+            LitKind::ByteStr(ref vec) => {
+                let vec_pat = self.next("vec");
+
+                println!("    if let LitKind::ByteStr(ref {}) = {}.node;", vec_pat, lit_pat);
+                println!("    if let [{:?}] = **{};", vec, vec_pat);
+            },
+            LitKind::Str(ref text, _) => {
+                let str_pat = self.next("s");
+
+                println!("    if let LitKind::Str(ref {}, _) = {}.node;", str_pat, lit_pat);
+                println!("    if {}.as_str() == {:?}", str_pat, &*text.as_str());
+            },
+        }
+    }
+
+    fn print_loop(&mut self, body: &Block<'_>, label: Option<Label>, des: LoopSource) {
+        if_chain! {
+            if let LoopSource::While = des;
+            if let Some(body_expr) = body.expr;
+            if let ExprKind::If(if_expr, if_body, _) = body_expr.kind;
+            then {
+                match if_expr.kind {
+                    ExprKind::DropTemps(cond) => {
+                        let condition_pat = self.next("condition");
+                        let body_pat = self.next("body");
+
+                        println!("    if let Some(While {{ {} @ condition, {} @ body }}) = higher::While::hir({})", condition_pat, body_pat, self.current);
+
+                        self.current = condition_pat;
+                        self.visit_expr(cond);
+
+                        self.current = body_pat;
+                        self.visit_expr(body_expr);
+                    },
+                    ExprKind::Let(pat, expr, _) => {
+                        let let_pat = self.next("let_pat");
+                        let let_expr_pat = self.next("let_expr");
+                        let if_then_pat = self.next("if_then");
+
+                        println!("    if let Some(WhileLet {{ {} @ let_pat, {} @ let_expr, {} @ if_then }}) = higher::WhileLet::hir({})", let_pat, let_expr_pat, if_then_pat, self.current);
```

Comment:
Ideally each of the `higher` outputs correspond to actually using `higher` in the `author` implementation. We shouldn't re-implement `higher`, but instead do `if let Some(...) = higher...` for every thing on every expression.

Although you can make an exception for `if` and `if let` since these are very simple and are not really desugaring. I wouldn't mind not even using `higher` for those.

One thing you could do is `match expr.span.ctxt().outer_expn_data().kind` to see what is desugared before invoking `higher`, but not necessary.